cephfs/file.go

```diff
@@ -0,0 +1,143 @@
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+	ret := C.ceph_open(mount.mount, cPath, C.int(flags), C.mode_t(mode))
+	if ret < 0 {
+		return nil, getError(ret)
+	}
+	return &File{mount: mount, fd: ret}, nil
+}
+
+// Close the file.
+//
+// Implements:
+//  int ceph_close(struct ceph_mount_info *cmount, int fd);
+func (f *File) Close() error {
+	return getError(C.ceph_close(f.mount.mount, f.fd))
+}
+
+// read directly wraps the ceph_read call. Because read is such a common
+// operation we deviate from the ceph naming and expose Read and ReadAt
+// wrappers for external callers of the library.
+//
+// Implements:
+//  int ceph_read(struct ceph_mount_info *cmount, int fd, char *buf, int64_t size, int64_t offset);
+func (f *File) read(buf []byte, offset int64) (int, error) {
+	bufptr := (*C.char)(unsafe.Pointer(&buf[0]))
+	ret := C.ceph_read(
+		f.mount.mount, f.fd, bufptr, C.int64_t(len(buf)), C.int64_t(offset))
+	if ret < 0 {
+		return int(ret), getError(ret)
+	}
+	return int(ret), nil
+}
+
+// Read data from file. Up to len(buf) bytes will be read from the file.
+// The number of bytes read will be returned.
+func (f *File) Read(buf []byte) (int, error) {
+	// to-consider: should we mimic Go's behavior of returning an
+	// io.ErrShortWrite error if write length < buf size?
+	return f.read(buf, -1)
+}
+
+// ReadAt will read data from the file starting at the given offset.
+// Up to len(buf) bytes will be read from the file.
+// The number of bytes read will be returned.
+func (f *File) ReadAt(buf []byte, offset int64) (int, error) {
+	return f.read(buf, offset)
+}
+
+// write directly wraps the ceph_write call. Because write is such a common
+// operation we deviate from the ceph naming and expose Write and WriteAt
+// wrappers for external callers of the library.
+//
+// Implements:
+//  int ceph_write(struct ceph_mount_info *cmount, int fd, const char *buf,
+//                 int64_t size, int64_t offset);
+func (f *File) write(buf []byte, offset int64) (int, error) {
+	bufptr := (*C.char)(unsafe.Pointer(&buf[0]))
+	ret := C.ceph_write(
+		f.mount.mount, f.fd, bufptr, C.int64_t(len(buf)), C.int64_t(offset))
+	if ret < 0 {
+		return 0, getError(ret)
+	}
+	return int(ret), nil
+}
+
+// Write data from buf to the file.
+// The number of bytes written is returned.
+func (f *File) Write(buf []byte) (int, error) {
+	return f.write(buf, -1)
+}
+
+// WriteAt writes data from buf to the file at the specified offset.
+// The number of bytes written is returned.
+func (f *File) WriteAt(buf []byte, offset int64) (int, error) {
+	return f.write(buf, offset)
+}
+
+// Seek will reposition the file stream based on the given offset.
+//
+// Implements:
+//  int64_t ceph_lseek(struct ceph_mount_info *cmount, int fd, int64_t offset, int whence);
+func (f *File) Seek(offset int64, whence int) (int64, error) {
+	// validate the seek whence value in case the caller skews
+	// from the seek values we technically support from C as documented.
```

Comment:
One reason I put this in was that I could not quickly tell if they were or not, in addition to wanting to ensure we throw an error for other never-valid ints.
Would you object to going with this "strict" form for now and then filing an issue to research if cephfs supports these and doing a follow up pr later to allow them if supported?

Reply:
I would suggest to include support for both. If CephFS does not support them, it should return a suitable error (`ENOSYS` or `ENOTSUPP`?). It may as well depend on the version of Ceph used, or the features enabled on the backend.

Reply:
While I was looking at the sources to verify the ceph_read behavior I also looked at seek.
It turns out that SEEK_DATA/SEEK_HOLE appear to be supported in mimic and onward. Unfortunately, on casual examination it also looks like mimic may abort if an invalid whence value is passed making it difficult to simply pass the values on to ceph and let it handle it.
Thus, I'm OK with adding the additional constants but I'd still like to defer the work to a future PR if you don't mind, just so we don't have to immediately add the version complexity in this initial patch series.

Reply:
Ok, in that case please add a small TODO comment about it.

Reply:
+1. I will add a TODO and file an issue.